cpp/dolfinx/la/MatrixCSR.h

```diff
@@ -0,0 +1,415 @@
+      std::copy(std::next(_data.data(), _row_ptr[local_size0 + i]),
+                std::next(_data.data(), _row_ptr[local_size0 + i + 1]),
+                std::next(ghost_value_data.begin(), val_pos));
+
+      insert_pos[neighbor_rank]
+          += _row_ptr[local_size0 + i + 1] - _row_ptr[local_size0 + i];
+    }
+
+    _ghost_value_data_in.resize(_val_recv_disp.back());
+
+    // Compute data sizes for send and receive from displacements
+    std::vector<int> val_send_count(_val_send_disp.size() - 1);
+    std::adjacent_difference(std::next(_val_send_disp.begin()),
+                             _val_send_disp.end(), val_send_count.begin());
+
+    std::vector<int> val_recv_count(_val_recv_disp.size() - 1);
+    std::adjacent_difference(std::next(_val_recv_disp.begin()),
+                             _val_recv_disp.end(), val_recv_count.begin());
+
+    int status = MPI_Ineighbor_alltoallv(
+        ghost_value_data.data(), val_send_count.data(), _val_send_disp.data(),
+        dolfinx::MPI::mpi_type<T>(), _ghost_value_data_in.data(),
+        val_recv_count.data(), _val_recv_disp.data(),
+        dolfinx::MPI::mpi_type<T>(), _neighbor_comm, &_request);
+    assert(status == MPI_SUCCESS);
+  }
+
+  /// Complete transfer of ghost rows to owning processes
+  /// Must be preceded by finalize_begin()
+  void finalize_end()
+  {
+    int status = MPI_Wait(&_request, MPI_STATUS_IGNORE);
+    assert(status == MPI_SUCCESS);
+
+    // Add to local rows
+    assert(_ghost_value_data_in.size() == _unpack_pos.size());
+    for (std::size_t i = 0; i < _ghost_value_data_in.size(); ++i)
+      _data[_unpack_pos[i]] += _ghost_value_data_in[i];
+
+    // Clear cache
+    const std::int32_t local_size0 = _index_maps[0]->size_local();
+    std::fill(std::next(_data.begin(), _row_ptr[local_size0]), _data.end(), 0);
+  }
+
+  /// Index maps for the row and column space. The row IndexMap contains ghost
+  /// entries for rows which may be inserted into and the column IndexMap
+  /// contains all local and ghost columns that may exist in the owned rows.
+  ///
+  /// @return Row and column index maps
+  const std::array<std::shared_ptr<const common::IndexMap>, 2>&
+  index_maps() const
+  {
+    return _index_maps;
+  }
+
+  /// Get local values
+  xtl::span<T> values() { return xtl::span(_data); }
+
+  /// Get local row pointers
+  xtl::span<std::int32_t> row_ptr() { return xtl::span(_row_ptr); }
+
+  /// Get local column indices
+  xtl::span<std::int32_t> cols() { return xtl::span(_cols); }
+
+private:
+  // Map describing the data layout for rows and columns
+  // including ghost rows and ghost columns
+  std::array<std::shared_ptr<const common::IndexMap>, 2> _index_maps;
+
+  // // Block size
+  std::array<int, 2> _bs;
+
+  // Data
+  std::vector<T, Allocator> _data;
+  std::vector<std::int32_t> _cols, _row_ptr;
+
+  // Precomputed data for finalize/update
+  // Neighborhood communicator
+  MPI_Comm _neighbor_comm;
```

Comment:
This should be a `dolfinx::MPI::Comm`. At the moment there is a leak because the communicator isn't being freed.

Reply:
Currently we have an issue with `MPI_Comm_dup` and MPICH for unweighted neighbor communicators.
Possible solution add weights to graphs, see #1944 .